### PR TITLE
Simple item transforming components, ports telebatons to transform components

### PR DESCRIPTION
**Pull request by silicons**
## Why It's Good For The Game

Components good, having to subtype everything under /melee/transforming bad.

## Changelog
:cl:
refactor: Added simple transforming components for items.
/:cl:

### CODE REVIEW

File: code/game/objects/items/melee/batons/classic.dm [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		H.spawn_gibs()
+		return (BRUTELOSS)
+
+/obj/item/melee/classic_baton/telescopic/proc/on_transform(obj/item/parent, new_active, mob/user, forced, datum/signal_source, checks_passed)
+	if(user)
+		if(new_active)
+			to_chat(user, "<span class ='warning'>You extend the baton.</span>")
+		else
+			to_chat(user, "<span class ='notice'>You collapse the baton.</span>")
+	if(new_active)
```
Comment (silicons) [inline code/game/objects/items/melee/batons/classic.dm:110]: i'm not sure what i should do here

maybe i can make a subtype of /transforming component that auto sets these i guess if we uh, really want to reduce copy paste

ugh, i fucking hate designing sometimes.

File: code/game/objects/items/melee/batons/classic.dm [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	if(!(SEND_SIGNAL(src, COMSIG_IS_TRANSFORMED) & COMPONENT_IS_TRANSFORMED))
+		return FALSE
+	return ..()
+
+/obj/item/melee/classic_baton/telescopic/suicide_act(mob/user)
+	var/mob/living/carbon/human/H = user
+	var/obj/item/organ/brain/B = H.getorgan(/obj/item/organ/brain)
+
+	user.visible_message("<span class='suicide'>[user] stuffs [src] up [user.p_their()] nose and presses the 'extend' button! It looks like [user.p_theyre()] trying to clear [user.p_their()] mind.</span>")
+	SEND_SIGNAL(src, COMSIG_TRANSFORM_ACTIVATE, user)
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm:93]: Better make this a suicide_act signal sent before suicide_act is called.

File: code/game/objects/items/melee/batons/classic.dm (hunk 1 above)
Comment (Ghommie) [reply code/game/objects/items/melee/batons/classic.dm:110]: Considering this transforming component is going to affect item size and combat values most times, yea, this should be shoved to the component itself.

File: code/game/objects/items/melee/batons/classic.dm [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		w_class = WEIGHT_CLASS_SMALL
+		force = 0 //not so robust now
+		attack_verb = list("prodded", "poked")
+	update_icon()
+	playsound(src.loc, 'sound/weapons/batonextend.ogg', 50, 1)
+	add_fingerprint(user)
+
+/obj/item/melee/classic_baton/update_icon()
+	. = ..()
+	if(SEND_SIGNAL(src, COMSIG_IS_TRANSFORMED) & COMPONENT_IS_TRANSFORMED)
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm:126]: update_icon_state and update_overlays signals exist and this whole function should be moved to the component itself, and yes, I know I still got to port some parent calls etcetera.

File: code/game/objects/items/melee/batons/classic.dm [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	item_flags = NONE
+	force = 0
+	total_mass = TOTAL_MASS_NORMAL_ITEM
+
+/obj/item/melee/classic_baton/telescopic/ComponentInitialize()
+	. = ..()
+	AddComponent(/datum/component/transforming, ATTACK_SELF_TRANSFORM, .proc/on_transform)
+
+/obj/item/melee/classic_baton/baton_stun(mob/living/target, mob/living/user)
+	if(!(SEND_SIGNAL(src, COMSIG_IS_TRANSFORMED) & COMPONENT_IS_TRANSFORMED))
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm]: Use `GetComponent` and check the `active` variable instead. The less snowflake signals, the better.

File: code/game/objects/items/melee/batons/classic.dm [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	var/stam_dmg = 30
+
+/obj/item/melee/classic_baton/attack(mob/living/target, mob/living/user)
+	if(!baton_stun(target, user))
+		return ..()
+
+// return true to disrupt normal melee attack chain
+/obj/item/melee/classic_baton/proc/baton_stun(mob/living/target, mob/living/user)
+	if(user.getStaminaLoss() >= STAMINA_SOFTCRIT)//CIT CHANGE - makes batons unusuable in stamina softcrit
+		to_chat(user, "<span class='warning'>You're too exhausted for that.</span>")//CIT CHANGE - ditto
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm:26]: `obj/item/attack `already says this line. And maybe remove the CIT CHANGE comments.

File: code/game/objects/items/melee/batons/classic.dm [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		if(ishuman(user))
+			var/mob/living/carbon/human/H = user
+			H.apply_damage(2*force, BRUTE, BODY_ZONE_HEAD)
+		else
+			user.take_bodypart_damage(2*force)
+		return TRUE
+	if(iscyborg(target))
+		return FALSE
+	if(!isliving(target))
+		return FALSE
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm]: The target should always be a living. So remove this tile.

File: code/game/objects/items/melee/batons/classic.dm [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	if (user.a_intent == INTENT_HARM)
+		return FALSE
+	else
```
Comment (Ghommie) [inline code/game/objects/items/melee/batons/classic.dm:60]: Remove this `else` that only makes things more indentated, the previous if statements lead to early returns anyway.

File: code/__DEFINES/components.dm [hunk 8]
```diff
@@ -330,6 +330,20 @@
...
+	#define COMPONENT_IS_TRANSFORMED				(1<<0)			//transform is active
+#define COMSIG_CAN_TRANSFORM					"can_transform"					//(mob/user) Invokes the usual checks, return TRUE/FALSE based on if it can transform.
+	#define COMPONENT_CAN_TRANSFORM					(1<<0)			//checks passed
+#define COMSIG_SET_TRANSFORM					"set_transform"					//(active, mob/user, force) - tries to set transform active to active. force will ignore can transform. returns true/false based on success/fail.
+#define COMSIG_TOGGLE_TRANSFORM					"toggle_transform"				//(mob/user, force) same as above but toggles
+#define COMSIG_TRANSFORM_ACTIVATE				"activate_transform"			//(mob/user, force) same as above but turns on. returns true if already on!
+#define COMSIG_TRANSFORM_DEACTIVATE				"deactivate_transform"			//(mob/user, force) same as above but turns off.
+	#define COMPONENT_TRANSFORM_SUCCESS				(1<<0)			//for the above 4 signals, means transforming was successful.
+	#define COMPONENT_ALREADY_TRANSFORMED			(1<<1)			//unnecessary, active is already what the signal wanted it to be
+
```
Comment (Ghommie) [inline code/__DEFINES/components.dm:349]: And remember, the less snowflake signals the better.

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/game/objects/items/melee/batons/classic.dm (hunk 5 above)
Comment (silicons) [reply code/game/objects/items/melee/batons/classic.dm:26]: attack isn't called, batons all have attack snowflaked. this needs to be here unless we want to make batons start using afterattack for whatever reason instead (which also isn't feasible)

File: code/game/objects/items/melee/batons/classic.dm (hunk 2 above)
Comment (silicons) [reply code/game/objects/items/melee/batons/classic.dm:93]: there's no suicide signals is there?

**Comment by silicons** [comment]
after further thought uh

this is probably a bad idea

if anyone wants to convince me otherwise do ping me, I am willing to continue work on this.

My reasoning is that unlike twohanding where it's actually a pain in the arse to get to work viacopypaste, doing an attack_hand active = !active and then on_transform() on different items is significantly less terrible + tedious. The memory usage that would be incurred from me making variables mirrored to components is just not worth it at this point tbh, and without the variables it's still the same amount of code/copypaste, just less memory intensive and at that point what's even the point of a component vs writing 5 lines of attack_self vs 1 add component call.

I'll probably port comsig updates and twohanding comopnents if Ghom doesn't do it though.

**Pull request closed without merging**